core/services/workflows/store/store_db.go

```diff
@@ -0,0 +1,331 @@
+
+	if state.Outputs.Value != nil {
+		p := values.Proto(state.Outputs.Value).GetMapValue()
+		ob, err := proto.Marshal(p)
+		if err != nil {
+			return workflowStepRow{}, err
+		}
+
+		wsr.OutputValue = ob
+	}
+
+	if state.Outputs.Err != nil {
+		errs := state.Outputs.Err.Error()
+		wsr.OutputErr = &errs
+	}
+
+	return wsr, nil
+}
+
+// `Add` creates the relevant workflow_execution and workflow_step entries
+// to persist the passed in ExecutionState.
+func (d *DBStore) Add(ctx context.Context, state *ExecutionState) error {
+	return d.transact(ctx, func(db *DBStore) error {
+		wex := &workflowExecutionRow{
+			ID:         state.ExecutionID,
+			WorkflowID: state.WorkflowID,
+			Status:     state.Status,
+		}
+		err := db.insertWorkflowExecution(ctx, wex)
+		if err != nil {
+			return fmt.Errorf("could not insert workflow execution %s: %w", state.ExecutionID, err)
+		}
+
+		ws := []workflowStepRow{}
+		for _, step := range state.Steps {
+			step, err := stateToStep(step)
+			if err != nil {
+				return err
+			}
+			ws = append(ws, step)
+		}
+		if len(ws) > 0 {
+			return db.upsertSteps(ctx, ws)
+		}
+		return nil
+	})
+}
+
+func (d *DBStore) upsertSteps(ctx context.Context, steps []workflowStepRow) error {
+	sql := `
+	INSERT INTO
+	workflow_steps(workflow_execution_id, ref, status, inputs, output_err, output_value)
+	VALUES (:workflow_execution_id, :ref, :status, :inputs, :output_err, :output_value)
+	ON CONFLICT ON CONSTRAINT uniq_workflow_execution_id_ref
+	DO UPDATE SET
+		workflow_execution_id = EXCLUDED.workflow_execution_id,
+		ref = EXCLUDED.ref,
+		status = EXCLUDED.status,
+		inputs = EXCLUDED.inputs,
+		output_err = EXCLUDED.output_err,
+		output_value = EXCLUDED.output_value;
+	`
+	stmt, args, err := sqlx.Named(sql, steps)
+	if err != nil {
+		return err
+	}
+	stmt = d.db.Rebind(stmt)
+	_, err = d.db.ExecContext(ctx, stmt, args...)
+	return err
+}
+
+func (d *DBStore) insertWorkflowExecution(ctx context.Context, execution *workflowExecutionRow) error {
+	sql := `
+	INSERT INTO
+	workflow_executions(id, workflow_id, status)
+	VALUES ($1, $2, $3)
+	`
+	_, err := d.db.ExecContext(ctx, sql, execution.ID, execution.WorkflowID, execution.Status)
+	return err
+}
+
+func (d *DBStore) transact(ctx context.Context, fn func(*DBStore) error) error {
+	return sqlutil.Transact(
+		ctx,
+		func(ds sqlutil.DataSource) *DBStore {
+			return &DBStore{db: ds}
+		},
+		d.db,
+		nil,
+		fn,
+	)
+}
+
+func (d *DBStore) GetUnfinished(ctx context.Context) ([]ExecutionState, error) {
+	sql := `
+	SELECT
+		workflow_steps.workflow_execution_id AS ws_workflow_execution_id,
+		workflow_steps.ref AS ws_ref,
+		workflow_steps.status AS ws_status,
+		workflow_steps.inputs AS ws_inputs,
+		workflow_steps.output_err AS ws_output_err,
+		workflow_steps.output_value AS ws_output_value,
+		workflow_executions.id AS we_id,
+		workflow_executions.workflow_id AS we_workflow_id,
+		workflow_executions.status AS we_status
+	FROM workflow_executions
+	JOIN workflow_steps
```

Comment:
nit: this join is a little wasteful because it adds execution state fields to every single step. It almost feels like you don't need to store execution state at all because you can infer the status from the set of steps + you could add workflowID to the step table. Not the greatest suggestion but I'm trying to think how to simplify the join ;)

Reply:
We could create a execution state view instead of having a concrete table for semantics sake

Reply:
I would rather keep the extra table at this stage... I think we *may* be able to dedupe right now, but I don't think this will be the case long-term as we're likely to add more information to the workflow execution.

I can avoid the join by making two queries and doing an in-memory join, but that felt worse to me. Wdyt?

Reply:
You could store the query as a view instead, then wholesale select the view from application code. That way if we ever end up dropping the extra table later on, you won't have to update application code, just the view.

I'd imagine it's best to avoid application level joins as much as possible?

Reply:
@HenryNguyen5 Doesn't that result in the same thing? If I understand what you're suggesting, we'd do two SELECTS and then an in-memory join?